### ISSUE TITLE
DofMap::local_variable_indices appears incorrect

### ISSUE THREAD

**Issue by YaqiWang** [libMesh/libmesh#1841]
I am running a case with multiple processors and with *block restricted* variables in LAGRANGE FIRST. Unfortunately the case is not in just libMesh so I cannot post it here. The size of indices for one variable obtained with this function is drastically smaller than the size with `System::local_dof_indices` (which returns a set) (5 vs 281 on master processor). I am guessing the logic at https://github.com/libMesh/libmesh/blob/master/src/base/dof_map.C#L1109 is not sufficient to catch my case.

I attempted a fix like this:
```
diff --git a/src/base/dof_map.C b/src/base/dof_map.C
index 5e259cc19..eac3d4551 100644
--- a/src/base/dof_map.C
+++ b/src/base/dof_map.C
@@ -1076,6 +1076,8 @@ void DofMap::local_variable_indices(std::vector<dof_id_type> & idx,
 {
   const unsigned int sys_num       = this->sys_number();
 
+  std::set<dof_id_type> temp;
+  unsigned int counter = 0;
   // If this isn't a SCALAR variable, we need to find all its field
   // dofs on the mesh
   if (this->variable_type(var_num).family != SCALAR)
@@ -1100,9 +1102,15 @@ void DofMap::local_variable_indices(std::vector<dof_id_type> & idx,
                   const dof_id_type index = node.dof_number(sys_num,var_num,i);
                   libmesh_assert_greater_equal (index, this->first_dof());
                   libmesh_assert_less (index, this->end_dof());
-
-                  if (idx.empty() || index > idx.back())
+                  temp.insert(index);
+                  if (temp.size() > counter)
+                  {
                     idx.push_back(index);
+                    ++counter;
+                  }
+
+//                  if (idx.empty() || index > idx.back())
+//                    idx.push_back(index);
                 }
             }
 
@@ -1111,8 +1119,14 @@ void DofMap::local_variable_indices(std::vector<dof_id_type> & idx,
           for (unsigned int i=0; i<n_comp; i++)
             {
               const dof_id_type index = elem->dof_number(sys_num,var_num,i);
-              if (idx.empty() || index > idx.back())
+              temp.insert(index);
+              if (temp.size() > counter)
+              {
                 idx.push_back(index);
+                ++counter;
+              }
+//              if (idx.empty() || index > idx.back())
+//                idx.push_back(index);
             }
         } // done looping over elements
 
@@ -1133,8 +1147,14 @@ void DofMap::local_variable_indices(std::vector<dof_id_type> & idx,
           for (unsigned int i=0; i<n_comp; i++)
             {
               const dof_id_type index = node->dof_number(sys_num,var_num,i);
-              if (idx.empty() || index > idx.back())
+              temp.insert(index);
+              if (temp.size() > counter)
+              {
                 idx.push_back(index);
+                ++counter;
+              }
+//              if (idx.empty() || index > idx.back())
+//                idx.push_back(index);
             }
         }
     }
```
It does fix my calculation, but the size of indices with this fix is larger than the size with `System::local_dof_indices` by 1. So in conclusion, I think this function is not correct for all cases and needs to be fixed. I do rely on the ordering of the indices in the vector, so using `System::local_dof_indices`, which orders indices by their value but not based on mesh topology, appears not an option for me. Thanks.

**Comment by YaqiWang** [comment]
If you think my patch is ok, I can send in a PR. I have hit this bug pretty frequently recently and need it to be fixed. Thanks.

**Comment by roystgnr** [comment]
Sorry, this completely slipped under my radar.  Can you set up a unit test that exhibits the off-by-1 problem remaining?

**Comment by YaqiWang** [comment]
No problem. You mean a unit test in libMesh? Never did that before, can you give a small instruction on how to do it?

**Comment by YaqiWang** [comment]
Dummy question, how do I run the test?

**Comment by roystgnr** [comment]
Assuming you have libMesh configured to use CppUnit (which we autodetect if cppunit-config is in your $PATH), ```make check -C tests``` from your libMesh build dir should do it.  It'll run all the *other* unit tests too, but we've been managing to keep them small so that won't take long.

**Comment by YaqiWang** [comment]
Working on a simple test, hopefully I can produce with it. Will let you know shortly.

**Comment by roystgnr** [comment]
Thanks!

**Comment by YaqiWang** [comment]
@roystgnr See the commit in this page. I did get one error on line 635, which is the first of my four tests. Hopefully this is enough.

**Comment by roystgnr** [comment]
Thanks!  Trying it (well, honestly, letting the computer try it while I head home for the evening) now.

**Comment by YaqiWang** [comment]
I feel it is a bit strange, cause I applied my patch and it does not change anything. Even I intentionally change the const in the assertions, the test seems always report one failure in the test.

**Comment by YaqiWang** [comment]
Sorry, my bad, I found one bug, should be 0.5 instead of 2 when setting subdomain IDs.

**Comment by YaqiWang** [comment]
Now I can see that Line 636 fails. Also my patch fixes the failure. Sorry, @roystgnr you need to re-pull the branch.

**Comment by roystgnr** [comment]
I'm starting to get it!

```DofMap::distribute_dofs_*``` iterates over elements, then over nodes on each element, skipping elements on which the variable currently being indexed is not supported because iff they support that variable from a different element then we'll hit it from that element later.

```DofMap::local_variable_indices``` iterates over elements, then over nodes on each element, *not* skipping elements based on subdomain (because we can just check the nodes right away to see if the variable is there) *but* assuming that the indices we see are going to be in ascending order.  With subdomain-specific variables that assumption breaks half the time with nodes on subdomain boundaries.

Let me try adding the subdomain test in local_variable_indices, as well as some screaming comments in both places to avoid repeating the bug if we change some numbering, and see if that gets it.

**Comment by roystgnr** [comment]
And the problem with System::local_dof_indices() crops up in parallel: when we have nodes that are owned by one processor but support subdomain-restricted variables for subdomains that aren't on the owning processor, then looping over elements without nodes misses their dofs.

**Comment by YaqiWang** [comment]
How can we design a unit test for both parallel and serial?

**Comment by roystgnr** [comment]
When LIBMESH_RUN='mpirun -np 4' (or some such), "make check" runs in parallel.

Your test gave false positives in parallel, but that was a 4-line fix.  I'll have a PR with my version of your branch shortly.

**Comment by YaqiWang** [comment]
This should be closed by #1877.
